internal/mage/util/engine.go

```diff
@@ -0,0 +1,312 @@
+	# enable controllers
+	sed -e 's/ / +/g' -e 's/^/+/' < /sys/fs/cgroup/cgroup.controllers \
+		> /sys/fs/cgroup/cgroup.subtree_control
+fi
+
+exec {{.EngineBin}} --config {{.EngineConfig}} {{ range $key := .EntrypointArgKeys -}}--{{ $key }}="{{ index $.EntrypointArgs $key }}" {{ end -}} "$@"
+`
+
+const engineConfigTmpl = `
+debug = true
+insecure-entitlements = ["security.insecure"]
+{{ range $key := .ConfigKeys }}
+[{{ $key }}]
+{{ index $.ConfigEntries $key }}
+{{ end -}}
+`
+
+var publishedEngineArches = []string{"amd64", "arm64"}
+
+// DevEngineOpts are options for the dev engine
+type DevEngineOpts struct {
+	EntrypointArgs map[string]string
+	ConfigEntries  map[string]string
+}
+
+func getEntrypoint(opts ...DevEngineOpts) (string, error) {
+	mergedOpts := map[string]string{}
+	for _, opt := range opts {
+		maps.Copy(mergedOpts, opt.EntrypointArgs)
+	}
+	keys := maps.Keys(mergedOpts)
+	sort.Strings(keys)
+
+	var entrypoint string
+
+	type entrypointTmplParams struct {
+		Bridge            string
+		EngineBin         string
+		EngineConfig      string
+		EntrypointArgs    map[string]string
+		EntrypointArgKeys []string
+	}
+	tmpl := template.Must(template.New("entrypoint").Parse(engineEntrypointTmpl))
+	buf := new(bytes.Buffer)
+	err := tmpl.Execute(buf, entrypointTmplParams{
+		EngineBin:         "/usr/local/bin/" + engineBinName,
+		EngineConfig:      engineTomlPath,
+		EntrypointArgs:    mergedOpts,
+		EntrypointArgKeys: keys,
+	})
+	if err != nil {
+		panic(err)
+	}
+	entrypoint = buf.String()
+
+	return entrypoint, nil
+}
+
+func getConfig(opts ...DevEngineOpts) (string, error) {
+	mergedOpts := map[string]string{}
+	for _, opt := range opts {
+		maps.Copy(mergedOpts, opt.ConfigEntries)
+	}
+	keys := maps.Keys(mergedOpts)
+	sort.Strings(keys)
+
+	var config string
+
+	type configTmplParams struct {
+		ConfigEntries map[string]string
+		ConfigKeys    []string
+	}
+	tmpl := template.Must(template.New("config").Parse(engineConfigTmpl))
+	buf := new(bytes.Buffer)
+	err := tmpl.Execute(buf, configTmplParams{
+		ConfigEntries: mergedOpts,
+		ConfigKeys:    keys,
+	})
+	if err != nil {
+		panic(err)
+	}
+	config = buf.String()
+
+	return config, nil
+}
+
+func CIDevEngineContainerAndEndpoint(ctx context.Context, c *dagger.Client, opts ...DevEngineOpts) (*dagger.Container, string, error) {
+	devEngine := CIDevEngineContainer(c, opts...)
+
+	endpoint, err := devEngine.Endpoint(ctx, dagger.ContainerEndpointOpts{Port: 1234, Scheme: "tcp"})
+	if err != nil {
+		return nil, "", err
+	}
+	return devEngine, endpoint, nil
+}
+
+var DefaultDevEngineOpts = DevEngineOpts{
+	EntrypointArgs: map[string]string{
+		"network-name": "dagger-dev",
+		"network-cidr": "10.88.0.0/16",
+	},
+	ConfigEntries: map[string]string{
+		"grpc":                 `address=["unix:///var/run/buildkit/buildkitd.sock", "tcp://0.0.0.0:1234"]`,
+		`registry."docker.io"`: `mirrors = ["mirror.gcr.io"]`,
```

Comment:
Oh nice! Does this have a big impact on the rate limits?

Reply:
@vito Yeah, it stops them from being an issue. 😄